Content.Server/Botany/Systems/MutationSystem.cs

```diff
@@ -0,0 +1,211 @@
+
+        result.Chemicals = random(0.5f) ? a.Chemicals : result.Chemicals;
+
+        CrossFloat(ref result.NutrientConsumption, a.NutrientConsumption);
+        CrossFloat(ref result.WaterConsumption, a.WaterConsumption);
+        CrossFloat(ref result.IdealHeat, a.IdealHeat);
+        CrossFloat(ref result.HeatTolerance, a.HeatTolerance);
+        CrossFloat(ref result.IdealLight, a.IdealLight);
+        CrossFloat(ref result.LightTolerance, a.LightTolerance);
+        CrossFloat(ref result.ToxinsTolerance, a.ToxinsTolerance);
+        CrossFloat(ref result.LowPressureTolerance, a.LowPressureTolerance);
+        CrossFloat(ref result.HighPressureTolerance, a.HighPressureTolerance);
+        CrossFloat(ref result.PestTolerance, a.PestTolerance);
+        CrossFloat(ref result.WeedTolerance, a.WeedTolerance);
+
+        CrossFloat(ref result.Endurance, a.Endurance);
+        CrossInt(ref result.Yield, a.Yield);
+        CrossFloat(ref result.Lifespan, a.Lifespan);
+        CrossFloat(ref result.Maturation, a.Maturation);
+        CrossFloat(ref result.Production, a.Production);
+        CrossFloat(ref result.Potency, a.Potency);
+
+        CrossBool(ref result.Seedless, a.Seedless);
+        CrossBool(ref result.Viable, a.Viable);
+        CrossBool(ref result.Slip, a.Slip);
+        CrossBool(ref result.Sentient, a.Sentient);
+        CrossBool(ref result.Ligneous, a.Ligneous);
+        CrossBool(ref result.Bioluminescent, a.Bioluminescent);
+        result.BioluminescentColor = random(0.5f) ? a.BioluminescentColor : result.BioluminescentColor;
+
+        // Hybrids have a high chance of being seedless. Balances very
+        // effective hybrid crossings.
+        if (a.Name == result.Name && random(0.7f))
+        {
+            result.Seedless = true;
+        }
+
+        return result;
+    }
+
+    // Mutate reference 'val' between 'min' and 'max' by pretending the value
+    // is representable by a thermometer code with 'bits' number of bits and
+    // randomly flipping some of them.
+    //
+    // 'totalbits' and 'mult' are used only to calculate the probability that
+    // one bit gets flipped.
+    private static void MutateFloat(ref float val, float min, float max, int bits, int totalbits, float mult)
+    {
+        // Probability that a bit flip happens for this value.
+        float p = mult*bits/totalbits;
+        if (!random(p))
+        {
+            return;
+        }
+
+        // Starting number of bits that are high, between 0 and n.
+        int n = (int)Math.Round((val - min) / (max - min) * bits);
+
+        // Probability that the bit flip increases n.
+        float p_increase = 1-(float)n/bits;
+        int np;
+        if (random(p_increase))
+        {
+            np = n + 1;
+        }
+        else
+        {
+            np = n - 1;
+        }
+
+        // Set value based on mutated thermometer code.
+        float nval = (float)Math.Min(Math.Max((float)np/bits * (max - min) + min, min), max);
```

Comment:
does `MathF.Min` et al allow you to remove the float casts

Reply:
Probably. I didn't know that MathF exists. I'll make the change.

Reply:
The outer float cast goes away. The inner float cast has to stay to coerce 'np' from integer to float.